py3status/modules/external_script.py

```diff
@@ -52,20 +99,55 @@ class Py3status:
     cache_timeout = 15
     format = '{output}'
     localize = True
+    notifications = {}
     script_path = None
     strip_output = False
 
     def post_config_hook(self):
         if not self.script_path:
             raise Exception(STRING_ERROR)
 
+        self.button_refresh = 2
+        self.notification = {'normal': []}
+        if self.notifications:
```

Comment:
@lasers could you describe what is happening in this `if` block?

Reply:
`post_config_hook` checking to find out which notifications to use... so we use a dict/list of booelans instead of constantly testing the conditionals on each interval.

Reply:
I'm looking at this right now and I think it's a bit over-engineered, both external_script and battery_level (and potentially updates) only need notifications on click (battery has extra special case of showing notification on certain threshold), is there any module that currently shows notifications always or when changed?

Reply:
IIRC I think battery's implementation of showing notification on clicks is a bit silly. We likely could make it show up on the bar instead. Maybe with a toggle to show/hide what we want.

Current and changed is similar. I'm not even sure if it makes sense to make a `changed` state... because we technically could make `changed` placeholder so you do `\?if=changed {test}` in current state. Maybe better not to. Fleshing out a nice notification solution is hard though.

* `wwan` on changed.
* `github` on always (always failing). Could be in `post_config_hook` instead.
* `battery_level` is both click and changed.
`pomodoro` is changed when timer reached zero. Not sure how we would add into this as it's not `changed`.

Things to take from this is that we are only making notification states to allow different kind of notifications to happen in a module. I guess we could use `changed` here.

* `graphite` seems to be always.

Reply:
* `wwan` on changed - 👍 
* `github` is not always, it's a weird way of showing error, completely custom use of notifications and should not be generalized
* `battery_level` is click and on certain thresholds, this is again custom and is not "changed" (i.e. it will not show notification when changing from 81% to 80%).
* `pomodoro` yes, it's again not changed but some custom logic.

I'm not sure how much we will be able to generalize if every module has some special use of notifications...